### DISCUSSION TITLE
pyUPDI (with diode) - error during programming

### DISCUSSION THREAD

**Discussion by DanSkareda** [SpenceKonde/megaTinyCore#446] [Q&A]
Hi. I built a diode programmer as recommended in project DxCore (https://github.com/SpenceKonde/AVR-Guidance/blob/master/UPDI/jtag2updi.md):

Ideal:
internal resistor in adapter: 2.2k >= Ra
resistor on target:   2k >= Rt >= 100 (100-470 rec.)

```
--------------------                                 To Target device
                DTR|                                  __________________
    internal    Rx |--------------,------------------| UPDI---\/\/---------->
  Tx---/\/\/\---Tx |-------|<|---'          .--------| Gnd    470 ohm (100 ~ 1k)
    resistor    Vcc|---------------------------------| Vcc
  typ 1-2k      CTS|                     .`          |__________________
                Gnd|--------------------'             If you make a 3-pin connector, use this pinout
--------------------
```
If I load the attiny414 circuit with this pyUPDI programmer and select "serial-UPDI with 4.7k resistor or diode SLOW...", everything is fine. But if I choose the normal option from TURBO, the programming ends with an error:

Arduino <---> pymcuprog bridge by Quentin Bolsee and Spence Konde
Version 1.1.0 - May 2021
Using serial port COM6 at 230400 baud.
Target: attiny414
Set fuses: ['2:0x02', '6:0x04', '8:0x00']
Action: write
File: C:\Users\radek\AppData\Local\Temp\arduino_build_761162/Blink.ino.hex
Pinging device...
Ping response: 1E9222
Setting fuse 0x2=0x2
Writing literal values...
Verifying literal values...
Action took 0.02s
Setting fuse 0x6=0x4
Writing literal values...
Verifying literal values...
Action took 0.00s
Setting fuse 0x8=0x0
Writing literal values...
Verifying literal values...
Action took 0.02s
Finished writing fuses.
Chip/Bulk erase,
Memory type eeprom is conditionally erased (depending upon EESAVE fuse setting)
Memory type flash is always erased
Memory type internal_sram is always erased
Memory type lockbits is always erased
...
Erased.
Action took 0.02s
Writing from hex file...
Writing flash...
[====                                              ] 1/11Traceback (most recent call last):
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2/tools/prog.py", line 266, in <module>
    main()
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2/tools/prog.py", line 121, in main
    return_code = pymcuprog_basic(args, fuses_dict)
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2/tools/prog.py", line 240, in pymcuprog_basic
Nastala chyba při nahrávaní projektu.
    blocksize=args.blocksize)
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2/tools/prog.py", line 131, in run_pymcu_action
    status = func(backend, *args, args_pymcu)
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\tools\libs\pymcuprog\pymcuprog_main.py", line 341, in _action_write
    _write_memory_segments(backend, result, args.verify, blocksize=args.blocksize)
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\tools\libs\pymcuprog\pymcuprog_main.py", line 392, in _write_memory_segments
    backend.write_memory(segment.data, memory_name, segment.offset, blocksize=blocksize)
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\tools\libs\pymcuprog\backend.py", line 526, in write_memory
    self.programmer.write_memory(data=data, memory_name=memory_name, offset=offset_byte, blocksize=blocksize)
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\tools\libs\pymcuprog\programmer.py", line 190, in write_memory
    self.device_model.write(memory, offset, data, blocksize=blocksize)
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\tools\libs\pymcuprog\nvmserialupdi.py", line 175, in write
    self.avr.nvm.write_flash(offset_aligned, chunk, blocksize=-1, bulkwrite = bulk)
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\tools\libs\pymcuprog\serialupdi\nvm.py", line 117, in write_flash
    return self.write_nvm(address, data, use_word_access=True, blocksize=blocksize,  bulkwrite=bulkwrite)
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\tools\libs\pymcuprog\serialupdi\nvm.py", line 198, in write_nvm
    self.execute_nvm_command(nvmcommand)
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\tools\libs\pymcuprog\serialupdi\nvm.py", line 76, in execute_nvm_command
    return self.readwrite.write_byte(self.device.nvmctrl_address + constants.UPDI_NVMCTRL_CTRLA, command)
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\tools\libs\pymcuprog\serialupdi\readwrite.py", line 64, in write_byte
    return self.datalink.st(address, value)
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\tools\libs\pymcuprog\serialupdi\link.py", line 300, in st
    return self._st_data_phase([value & 0xFF])
  File "C:\Users\radek\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\tools\libs\pymcuprog\serialupdi\link.py", line 246, in _st_data_phase
    raise PymcuprogError("Error with st")
pymcuprog.pymcuprog_errors.PymcuprogError: Error with st

**Comment by hmeijdam** [comment]
I built mine from [this CH340 adapter](https://www.aliexpress.com/item/32833772712.html?spm=a2g0s.9042311.0.0.27424c4d2fIpQZ) without any resistors and only a diode between RX ->|- TX on the adapter and it is stable on all 3 speeds, including 460800 baud.

**Reply by DanSkareda** [reply]
I have built many pyUPDI adapters with CH340n, but they don't have a diode in the circuit. They work at all speeds. The problem is not that I can't build a problem-free pyUPDI programmer, but that this circuit is recommended by the author as the ideal circuit and it doesn't work - it has programming errors.

**Reply by DanSkareda** [reply]
I don't think a circuit without resistors in the circuit would be safe. In case of an accidental short circuit, the programmer could be damaged. That is why it is good to have some current limiting resistor in the circuit. The diode alone will probably work, but I don't think it's very safe for the programmer. But this solves the ideal circuit - 1k - 2k resistor on the programmer side, 470R on the programmer side. And of course the diode.

**Reply by hmeijdam** [reply]
I agree it's not the safest circuit, but bought them at 5pcs for 3 euro, so when I blow up the 3rd, I'll add some resistors :-)
[edit] I should mention, that I have only the ground and UPDI wire connected. I power the MCU with a dedicatd power supply.

**Reply by SpenceKonde** [reply]
What kind of diodes? What serial adapter are you using them with?  Whats the value of the on-board resistor on both ends?  (trying to get enough information that I can attempt to reproduce this. Mine all work perfectly, but am getting complaints from several people that my circuit doesn't work.

**Reply by SpenceKonde** [reply]
And does it work under the same circumstances with different electrical design (as opposed to the diode)? There is an issue that impacts some serial adapters regardless of the method used for that connection that can cause that sort of failure when programming tinyAVR devices at non SLOW speeds which needs to be fixed in software and is going to be tricky to do while minimizing the performance impact with serial adapters that work fine now. This one is caused by the serial adapter picking back up too fast after the page write command. most adapters have latency time longer than the page write!) Only see that behavior with the anomalously low-latency HT42B534 (and the issue is noted in that document),. And it is entirely possible that that's what is happening in that error.

**Reply by SpenceKonde** [reply]
Yeah the traceback looks like the software bug.

**Reply by hmeijdam** [reply]
While responding to my answer, I assume your question is for @DanSkareda ?
FWIW I am not using a schottkey diode but just grabbed some unknown signal diode with forward voltage of 0.7 volt.
![Serial programmer back](https://user-images.githubusercontent.com/9595465/120807007-c8f85580-c547-11eb-8098-cd2ec2aef65c.jpg)

**Reply by DanSkareda** [reply]
My programmer is this (http://imgway.cz/s/O70). My own production. There is no resistor in the target circuit at the UPDI input. They are only in the programmer. In a slightly different circuit (without diode and with different resistors) this pyUPDI works at all speeds, even TURBO.

**Comment by SpenceKonde** [comment]
These branching threads are wack, I don't like how threads work here... I wish it would just work like a normal forum. 

@hmeijdam I had problems with non-schottky diodes too,. thats' why I specify a small signal schottky diode.

**Comment by DanSkareda** [comment]
I built a second programmer and it's the same. Can you please fix this problem?

**Reply by SpenceKonde** [reply]
Interesting. I think I know what's happening.... I am having a call today or tomorrow with my co-author to discuss the steps we should take before making it an independently available release. I think the data about the following page is being sent before it finishes writing the previous page, and you have only 1 byte of buiffer in that case. And it works on most parts at least for me due to USB latency, and fails in cases where there isn't enough latency, the same reason that the HT42-based adapters don't work except on "SLOW"

**Reply by SpenceKonde** [reply]
I have a fix for this in and it will be available in the next release.

**Reply by DanSkareda** [reply]
Thanks. Then I'll try.
